crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs

```diff
@@ -0,0 +1,532 @@
+                return None;
+            }
+        }
+        // Step (2): parse the `__all__` definition using the raw tokens.
+        //
+        // (2a). Start by collecting information on each line individually:
+        let lines = collect_dunder_all_lines(*range, locator)?;
+
+        // (2b). Group lines together into sortable "items":
+        //   - Any "item" contains a single element of the `__all__` list/tuple
+        //   - "Items" are ordered according to the element they contain
+        //   - Assume that any comments on their own line are meant to be grouped
+        //     with the element immediately below them: if the element moves,
+        //     the comments above the element move with it.
+        //   - The same goes for any comments on the same line as an element:
+        //     if the element moves, the comment moves with it.
+        let items = collect_dunder_all_items(&lines);
+
+        Some(DunderAllValue {
+            items,
+            range: *range,
+            multiline: is_multiline,
+        })
+    }
+
+    fn construct_sorted_all(
+        &self,
+        locator: &Locator,
+        parent: &ast::Stmt,
+        stylist: &Stylist,
+    ) -> SortedDunderAll {
+        let mut sorted_items = self.items.clone();
+        sorted_items.sort();
+
+        // As well as saving us unnecessary work,
+        // returning early here also means that we can rely on the invariant
+        // throughout the rest of this function that both `items` and `sorted_items`
+        // have length of at least two. If there are fewer than two items in `__all__`,
+        // it is impossible for them *not* to compare equal here:
+        if sorted_items == self.items {
+            return SortedDunderAll::AlreadySorted;
+        }
+        assert!(self.items.len() >= 2);
+
+        // As well as the "items" in the `__all__` definition,
+        // there is also a "prelude" and a "postlude":
+        //  - Prelude == the region of source code from the opening parenthesis
+        //    (if there was one), up to the start of the first element in `__all__`.
+        //  - Postlude == the region of source code from the end of the last
+        //    element in `__all__` up to and including the closing parenthesis
+        //    (if there was one).
+        let prelude_end = {
+            // We should already have returned by now if there are 0 items:
+            // see earlier comments in this function
+            let first_item = self
+                .items
+                .first()
+                .expect("Expected there to be at least two items in the list");
+            let first_item_line_offset = locator.line_start(first_item.start());
+            if first_item_line_offset == locator.line_start(self.start()) {
+                first_item.start()
+            } else {
+                first_item_line_offset
+            }
+        };
+        let (needs_trailing_comma, postlude_start) = {
```

Comment:
Do I understand this correctly (without having read through all the code) that the fix inserts a trailing comma if the last item is on its own line? 

What about users that prefer not to have trailing commas? Should we instead detect if the source contains a trailing comma and preserve it?

Reply:
> Do I understand this correctly (without having read through all the code) that the fix inserts a trailing comma if the last item is on its own line?

yes, that's correct

> What about users that prefer not to have trailing commas? Should we instead detect if the source contains a trailing comma and preserve it?

good idea; I'll try to implement that

Reply:
Tweaked it in 6f5f55843c7bfc3ea3f7a336cacfceb33eeeb9c3: we now do not add a trailing comma if there wasn't one in the original source!